Refactor and types

ctwarog-ross opened the pull request:
Closes #1146
Closes #1139
Progress on #1151

http objects abstracted out of server endpoint callbacks implemented new error type catena.StatusResult for callback returns to signify operation result All non struct types are handled and provided in basic_device_bl.go

ctwarog-ross committed:
Refactor and types

Closes #1146
Closes #1139
Progress on #1151

http objects abstracted out of server endpoint callbacks
implemented new error type catena.StatusResult for callback returns to signify operation result
All non struct types are handled and provided in basic_device_bl.go

abrown-RV commented on sdks/go/pkg/catena/status_result.go:
```diff
@@ -0,0 +1,20 @@
+)
+// StatusResult represents an HTTP outcome (success or error) that the server can render uniformly.
+type StatusResult struct {
+    Status  int    // e.g., 200, 204, 400, 404, 405, 501, 500
```
This is still quite HTTP centric, fine for now, but when the time comes we'll have to decide how we map these constructors to the different http/grpc codes.

abrown-RV commented on sdks/go/pkg/catena/status_result.go:
```diff
@@ -0,0 +1,20 @@
+func NotFound(msg string) StatusResult            { return StatusResult{Status: http.StatusNotFound, Message: msg} }
+func MethodNotAllowed(msg string) StatusResult    { return StatusResult{Status: http.StatusMethodNotAllowed, Message: msg} }
+func NotImplemented(msg string) StatusResult      { return StatusResult{Status: http.StatusNotImplemented, Message: msg} }
+func InternalServerError(msg string) StatusResult { return StatusResult{Status: http.StatusInternalServerError, Message: msg} }
```
not totally up on go, but can the status result have private fields so that no one can monkey with it? is there a catena.StatusResult() ctor that would create an empty one? both things we should prevent if we can.

Basically I'm wanting to make sure you can't circumvent the Convenience constructors.

ctwarog-ross replied:
will be refactored in future work

abrown-RV commented on sdks/go/pkg/catena/status_result.go:
```diff
@@ -0,0 +1,20 @@
+type StatusResult struct {
+    Status  int    // e.g., 200, 204, 400, 404, 405, 501, 500
+    Payload any    // optional body to JSON-encode on success (2xx). If nil and Status is 204, no body.
+    Message string // optional error message for non-2xx statuses
```
Maybe want a Exception field as well? would let creators add exception details that could at least be logged, even if it wouldn't go back to the user. Although what's our pattern, do we just let callbacks throw exceptions?

ctwarog-ross replied:
Thats the Message field, I felt that name was more appropriate. Or are you thinking another string?

abrown-RV commented on sdks/go/pkg/rest/server.go:
```diff
@@ -1,226 +1,248 @@
+// Handlers now return HTTPResult so the server can respond consistently.
+type DeviceHandler func(w http.ResponseWriter, r *http.Request) catena.StatusResult
+type GetParamValueHandler func(slot int, fqoid string) catena.StatusResult
+type SetParamValueHandler func(value any, slot int, fqoid string) catena.StatusResult
```
cpp sdk just calls these GetValue and SetValue should we be consistant?

ctwarog-ross replied:
Im actually not sure, do you think we should?

abrown-RV commented on sdks/go/pkg/rest/server.go:
```diff
@@ -1,226 +1,248 @@
+                    return
+                }
+                writeHTTPResult(w, s.DefaultGetParamValueHandler(w, r, slot, fqoid))
             case http.MethodPut, http.MethodPatch:
```
why patch? openapi.yaml only specifies PUT and GET for /value

ctwarog-ross replied:
good catch, chatgpt put patch in there. It should be removed

abrown-RV commented on sdks/go/pkg/rest/server.go:
```diff
@@ -1,226 +1,248 @@
-				logger.Warning("%s /value/%s: method not allowed", r.Method, fqoid)
-                http.Error(w, "method not allowed", http.StatusMethodNotAllowed)
+                logger.Warning("%s /value/%s: method not allowed", r.Method, fqoid)
+                writeHTTPResult(w, catena.MethodNotAllowed("method not allowed"))
```
Method is pretty specific to HTTP, I know its convienent to use the StatusResult here, but I think MethodNotAllowed pollutes the StatusResult with a too specific condition. And we're not in bix logic code here, so it's ok to not use StatusResult

abrown-RV commented on sdks/go/pkg/rest/server.go:
```diff
@@ -1,226 +1,248 @@
+    logger.Debug("Registering routes for %d slots", len(s.slotList))
     for _, slot := range s.slotList {
         // capture loop variable
         slot := slot
```
vscode seems to think this is not needed?

ctwarog-ross replied:
slot is used several times in the function body

abrown-RV commented on sdks/go/pkg/rest/server.go:
```diff
@@ -1,226 +1,248 @@
 }
 
-func (s *Server) RegisterSetParamHandler(slot int, handler ParamValueHandler) {
+func (s *Server) RegisterSetParamHandler(slot int, handler SetParamValueHandler) {
```
`RegisterSetParamValueHandler` or `RegisterSetValueHandler`

abrown-RV commented on sdks/go/pkg/rest/server.go:
```diff
@@ -1,226 +1,248 @@
 }
 
-func (s *Server) RegisterGetParamHandler(slot int, handler ParamValueHandler) {
+func (s *Server) RegisterGetParamHandler(slot int, handler GetParamValueHandler) {
```
`RegisterGetParamValueHandler` or `RegisterGetValueHandler`

abrown-RV commented on sdks/go/examples/basic_device_bl/basic_device_bl.go:
```diff
@@ -1,100 +1,76 @@
 
-//For a given slot implement param handlers for a specific param oid only
+// For a given slot implement param handlers for a specific param oid only
 func registerSpecificParamHandlers(srv *rest.Server, params *sync.Map, fqoid string, slot int) {
```
Do we want this as an example? Isn't basically the same as the Basic Handler, and you would not be able to register 2 specific handlers for the same slot. As in you could not have
```go
registerSpecificParamHandlers(srv, params1, "alpha", 1)
registerSpecificParamHandlers(srv, params1, "counter", 1)
```
the specific handler for `alpha` would be overwritten by the handler for `counter` so it seems like a bad pattern at have as an example.

ctwarog-ross replied:
it was just to show you can implement more specific handlers than a blanket case

abrown-RV commented on sdks/go/examples/basic_device_bl/basic_device_bl.go:
```diff
@@ -1,100 +1,76 @@
-        params.Store(fqoid, v)
-        w.WriteHeader(http.StatusNoContent)
+        params.Store(fqoid, value)
+        return catena.NoContent()
```
I feel like we could be doing a little more for the return type... Here we are relying on the user to know that SetValue is supposed to return `NoContent` on success, but other endpoints return `OK` on success. I think we should just have one success type and the server handles converting that to the approriate response. Oh this is actually a grpc/rest argument as well, in grpc SetValue responds with an st2138.empty and http with 204 status.

abrown-RV reviewed (commented):
Just some comments, I know there's more changes in the pipe so I'll save approval for then

ctwarog-ross committed:
errors in WriteResponseJSON correctly handled
GetParamValue now GetValue
SetParamValue now SetValue
structs now implemented
PR comments addressed

NR-RV commented on sdks/go/examples/basic_device_bl/basic_device_bl.go:
```diff
@@ -2,160 +2,163 @@ package main
+            StructVariantType: "typeB",
+            Value:             123,
+        },
+    })
```
intenting switched from tabs to spaces

abrown-RV commented on sdks/go/pkg/rest/server.go:
```diff
@@ -6,221 +6,239 @@ import (
 	logger.Warning("PUT /value/%s: no handler registered for slot %d", fqoid, slot)
-	http.Error(w, "no setParamValue handler registered for slot "+strconv.Itoa(slot)+": fqoid /"+fqoid, http.StatusNotImplemented)
+	return catena.NotImplemented("no setParamValue handler registered for slot " + strconv.Itoa(slot))
 }
```
For consistancy should rename these too

ctwarog-ross replied:
wdym in progress logging? I have what was in his original submission

abrown-RV commented on sdks/go/pkg/rest/server.go:
```diff
@@ -6,221 +6,239 @@ import (
+	h, ok := s.getValue[slot]
+	return h, ok
 }
 
-func (s *Server) lookupSetParamValue(slot int) (ParamValueHandler, bool) {
-    s.mu.RLock()
-    defer s.mu.RUnlock()
-    if m, ok := s.setParamValue[slot]; ok {
-        return m, ok
-    }
-    return nil, false
+func (s *Server) lookupSetParamValue(slot int) (SetValueHandler, bool) {
+	s.mu.RLock()
+	defer s.mu.RUnlock()
+	h, ok := s.setValue[slot]
+	return h, ok
 }
```
for consistancy should rename these too

abrown-RV reviewed (changes requested):
You've got the in-progress logging stuff in here somehow

ctwarog-ross committed:
Applied pr feedback

abrown-RV commented on sdks/go/internal/json_utils.go:
```diff
@@ -13,43 +15,144 @@ func ReadValueFromRequest(r *http.Request) (any, error) {
-			if s, ok := stringVal.(string); ok {
-				return s, nil
+		}
+	} else if structVariantArrVal, ok := val["struct_variant_array_value"]; ok {
```
These should be recursive

abrown-RV commented on sdks/go/internal/json_utils.go:
```diff
@@ -13,43 +15,144 @@ func ReadValueFromRequest(r *http.Request) (any, error) {
+		for _, item := range val {
+			structVariantArr = append(structVariantArr, map[string]any{
+				"struct_variant_type": item.StructVariantType,
+				"value":              item.Value,
```
recursive here too, on all the struct types
